### PR TITLE
Migrate HttpClientTest test server to Armeria

### PR DESCRIPTION
**Pull request by anuraaga**
My main motivation for this is to follow up with migrating the HTTPS test from an external request to internal to work offline, and also add HTTP/2 tests for libraries that support it. I was sad when I found that Netty instrumentation tests wouldn't work without introducing shading, but in the end I think it's good given we've found some issues with Jetty conflicts between instrumentation and testing in #3079 too, I guess it's best practice to shade the server we use for tests.

If this PR is merged, I will follow up with removing remaining usage of `TestHttpServer` and delete it.

The build won't pass though, reactor-netty fails mysteriously - as far as I can tell the connection context is not getting cleaned up so only the first test passes, and all other client spans get suppressed after that. Armeria presumably has triggered some connection reuse behavior that wasn't there before, my attempt to try to prevent it with `maxRequestsPerCnonection(1)` didn't help. But I guess the instrumentation has some bug - @iNikem could you help troubleshoot?

### CODE REVIEW

File: instrumentation/akka-http-10.0/javaagent/akka-http-10.0-javaagent.gradle [hunk 1]
```diff
@@ -51,16 +45,9 @@ dependencies {
...
-  version101TestImplementation "com.typesafe.akka:akka-http_2.11:10.1.0"
-  version101TestImplementation "com.typesafe.akka:akka-stream_2.11:2.5.11"
-}
-
-test.dependsOn version101Test
-
-compileVersion101TestGroovy {
-  classpath = classpath.plus(files(compileVersion101TestScala.destinationDir))
-  dependsOn compileVersion101TestScala
+  // 10.0.x has some bugs making it incompatible with our test server.
```
Comment (anuraaga) [inline instrumentation/akka-http-10.0/javaagent/akka-http-10.0-javaagent.gradle:48]: I tried latest version of 10.0.x too but it still didn't sure. Not sure what the bug is but hopefully this is OK.

File: instrumentation/akka-http-10.0/javaagent/akka-http-10.0-javaagent.gradle [hunk 2]
```diff
@@ -51,16 +45,9 @@ dependencies {
...
   library "com.typesafe.akka:akka-stream_2.11:2.4.14"
 
-  // There are some internal API changes in 10.1 that we would like to test separately for
-  version101TestImplementation "com.typesafe.akka:akka-http_2.11:10.1.0"
-  version101TestImplementation "com.typesafe.akka:akka-stream_2.11:2.5.11"
-}
-
-test.dependsOn version101Test
-
-compileVersion101TestGroovy {
```
Comment (anuraaga) [inline instrumentation/akka-http-10.0/javaagent/akka-http-10.0-javaagent.gradle:61]: I couldn't find any difference between this pattern and the standard latestTestDeps, we generally are fine with it vs testing the baseline of a new minor I think.

Comment (anuraaga) [reply instrumentation/akka-http-10.0/javaagent/akka-http-10.0-javaagent.gradle:61]: Ah to clarify, even if I didn't have to change to `testLibrary` for the mysterious 10.0.x failure, I would have still made this change

File: instrumentation/java-http-client/javaagent/src/test/groovy/JdkHttpClientTest.groovy [hunk 3]
```diff
@@ -64,40 +62,4 @@ class JdkHttpClientTest extends HttpClientTest<HttpRequest> implements AgentTest
...
-          kind CLIENT
-          attributes {
-            "${SemanticAttributes.NET_TRANSPORT.key}" IP_TCP
-            "${SemanticAttributes.NET_PEER_NAME.key}" uri.host
-            "${SemanticAttributes.NET_PEER_IP.key}" { it == null || it == "127.0.0.1" }
-            // Optional
-            "${SemanticAttributes.NET_PEER_PORT.key}" uri.port > 0 ? uri.port : { it == null || it == 443 }
-            "${SemanticAttributes.HTTP_URL.key}" { it == "${uri}" || it == "${removeFragment(uri)}" }
-            "${SemanticAttributes.HTTP_METHOD.key}" method
-            "${SemanticAttributes.HTTP_FLAVOR.key}" "2.0"
```
Comment (anuraaga) [inline instrumentation/java-http-client/javaagent/src/test/groovy/JdkHttpClientTest.groovy:92]: This test was only here to override the base class for this assertion. I forced the client to HTTP1 instead and will add proper HTTP2 tests in the future.

**Review by iNikem — COMMENTED** [review]
What reactor-netty? :) I see other failed tests: https://scans.gradle.com/s/niu4tnujvnxsu/tests

**Comment by anuraaga** [comment]
@inikem Ah looks like a difference of localhost vs 127.0.0.1 caused by the execution environment difference, will tweak that. Still would help if you could try this branch's reactor-netty tests since those seem like real breakage

File: instrumentation/java-http-client/javaagent/src/test/groovy/JdkHttpClientTest.groovy (hunk 3 above)
Comment (anuraaga) [reply instrumentation/java-http-client/javaagent/src/test/groovy/JdkHttpClientTest.groovy:92]: Sure filed #3227

**Comment by iNikem** [comment]
> @iNikem Ah looks like a difference of localhost vs 127.0.0.1 caused by the execution environment difference, will tweak that. Still would help if you could try this branch's reactor-netty tests since those seem like real breakage (and reactor netty instrumentation is extremely hard to grok :(

Still don't see failing reactor-netty tests :D https://scans.gradle.com/s/iw4tlx733ld2g I see just netty failure

**Comment by anuraaga** [comment]
@iNikem Hmm, test (8) is a testcontainers random flake, build looks like a flakiness of some sort though will see if there is extra flakiness because of this PR.

But test (15) shows it (as does running locally every time :P)

https://scans.gradle.com/s/5naapz3zi7hrg

I think it's because the client span remains in the channel through the tests and couldn't figure out where that's supposed to get cleared out.

File: instrumentation/reactor-netty/reactor-netty-0.9/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/reactornetty/v0_9/AbstractReactorNettyHttpClientTest.groovy [hunk 4]
```diff
@@ -40,19 +42,27 @@ abstract class AbstractReactorNettyHttpClientTest extends HttpClientTest<HttpCli
...
+      .baseUrl(resolveAddress("").toString())
       ."${method.toLowerCase()}"()
       .uri(uri.toString())
   }
 
   @Override
   int sendRequest(HttpClient.ResponseReceiver request, String method, URI uri, Map<String, String> headers) {
-    return request.response().block().status().code()
+    return request.responseSingle {resp, content ->
+      // Make sure to consume content since that's when we close the span.
```
Comment (anuraaga) [inline instrumentation/reactor-netty/reactor-netty-0.9/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/reactornetty/v0_9/AbstractReactorNettyHttpClientTest.groovy:53]: This monstrosity is the only way I found to have reactor-netty present us a status code after consuming the response.

File: instrumentation/akka-http-10.0/javaagent/akka-http-10.0-javaagent.gradle (hunk 2 above)
Comment (anuraaga) [reply instrumentation/akka-http-10.0/javaagent/akka-http-10.0-javaagent.gradle:61]: Was able to fix issue by discarding entity

**Comment by anuraaga** [comment]
Looks like issues were related to Armeria flushing headers before content, causing tests to also see headers before content was consumed. Fixed the tests to consume explicitly

**Review by iNikem — APPROVED** [review]

**Pull request merged**